### ISSUE TITLE
ICE: glacier ./77329.rs  assertion failed: `(left == right)` DefId

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6251]
### Code

```rust
fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| { x } }]> {
    std::iter::empty()
}
```

### Meta

```
clippy 0.0.212 (ffa2e7a 2020-10-24)


rustc 1.46.0-nightly (f455e46ea 2020-06-20)
binary: rustc
commit-hash: f455e46eae1a227d735091091144601b467e1565
commit-date: 2020-06-20
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

### Error output

```
warning: unnecessary braces around const expression
 --> ./77329.rs:1:42
  |
1 | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| { x } }]> {
  |                                          ^^^^^^^^^^^^^^^^^^^ help: remove these braces
  |
  = note: `#[warn(unused_braces)]` on by default

thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `DefId(0:5 ~ 77329[317d]::bug::{opaque#0})`,
 right: `DefId(0:3 ~ 77329[317d]::bug)`', compiler/rustc_middle/src/hir/map/collector.rs:226:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.0.212 (ffa2e7a 2020-10-24)

query stack during panic:
#0 [index_hir] index HIR
#1 [hir_owner] HIR owner of `{misc#0}`
end of query stack
```

<details><summary>Backtrace</summary>
  <p>
  
  ```
  warning: unnecessary braces around const expression
 --> ./77329.rs:1:42
  |
1 | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| { x } }]> {
  |                                          ^^^^^^^^^^^^^^^^^^^ help: remove these braces
  |
  = note: `#[warn(unused_braces)]` on by default

thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `DefId(0:5 ~ 77329[317d]::bug::{opaque#0})`,
 right: `DefId(0:3 ~ 77329[317d]::bug)`', compiler/rustc_middle/src/hir/map/collector.rs:226:13
stack backtrace:
   0:     0x7fc83bf2aa20 - std::backtrace_rs::backtrace::libunwind::trace::h303a626d53553a64
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:100:5
   1:     0x7fc83bf2aa20 - std::backtrace_rs::backtrace::trace_unsynchronized::h6abb882733c3f18c
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fc83bf2aa20 - std::sys_common::backtrace::_print_fmt::h57829321f1d9217a
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7fc83bf2aa20 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h400b66a054640aac
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7fc83bf99e9c - core::fmt::write::h1857a60b204f1b6a
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/core/src/fmt/mod.rs:1076:17
   5:     0x7fc83bf1c642 - std::io::Write::write_fmt::h6453b28c8b27c6ea
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/io/mod.rs:1516:15
   6:     0x7fc83bf2f9d5 - std::sys_common::backtrace::_print::h15cc4b3923cc9bc0
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7fc83bf2f9d5 - std::sys_common::backtrace::print::h953c7be178c0674d
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7fc83bf2f9d5 - std::panicking::default_hook::{{closure}}::h635d667f571198f0
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:208:50
   9:     0x7fc83bf2f678 - std::panicking::default_hook::h3bc2efb48060abaa
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:227:9
  10:     0x55bcb3bae491 - clippy_driver::report_clippy_ice::h76ee1ad508e689ae
  11:     0x7fc83bf30216 - std::panicking::rust_panic_with_hook::h887eb0c89f46b8b2
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:581:17
  12:     0x7fc83bf2fd99 - std::panicking::begin_panic_handler::{{closure}}::hf0d3f46d37dd3bc1
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:484:9
  13:     0x7fc83bf2aebc - std::sys_common::backtrace::__rust_end_short_backtrace::h4bce8b52717576e3
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:141:18
  14:     0x7fc83bf2fd59 - rust_begin_unwind
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:483:5
  15:     0x7fc83bf2fd0b - std::panicking::begin_panic_fmt::h53297a84a5b19714
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:437:5
  16:     0x7fc83f642b8a - rustc_middle::hir::map::collector::NodeCollector::insert_entry::hbd98080ae7d33772
  17:     0x7fc83f642d09 - <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_param::ha8613785830c387d
  18:     0x7fc83f17a9aa - rustc_hir::intravisit::walk_body::h64fb9a7ff35c5c8d
  19:     0x7fc83f17bcac - rustc_hir::intravisit::walk_expr::h64dd9b0c405769b5
  20:     0x7fc83f17611f - rustc_hir::intravisit::walk_block::h7f89d11e0eb57d02
  21:     0x7fc83f17b21f - rustc_hir::intravisit::walk_expr::h64dd9b0c405769b5
  22:     0x7fc83f17aa0c - rustc_hir::intravisit::walk_body::h64fb9a7ff35c5c8d
  23:     0x7fc83f179cc2 - rustc_hir::intravisit::walk_ty::h3e03bb55739eab3b
  24:     0x7fc83f179109 - rustc_hir::intravisit::walk_assoc_type_binding::h6b229fc064e427d3
  25:     0x7fc83f17d67c - rustc_hir::intravisit::walk_path::hbb79bb99e067d6dc
  26:     0x7fc83f641890 - rustc_hir::intravisit::Visitor::visit_param_bound::hc426c8f8968a0549
  27:     0x7fc83f17c43f - rustc_hir::intravisit::walk_item::h3f2aa44f9a7cca2c
  28:     0x7fc83f642fd2 - <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_item::hcf243774ab523eaf
  29:     0x7fc83f179bc6 - rustc_hir::intravisit::walk_ty::h3e03bb55739eab3b
  30:     0x7fc83f17950a - rustc_hir::intravisit::walk_fn::h738d8fd4d8e7320a
  31:     0x7fc83f643a61 - <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_fn::hb414a3951be9470d
  32:     0x7fc83f17c667 - rustc_hir::intravisit::walk_item::h3f2aa44f9a7cca2c
  33:     0x7fc83f642fd2 - <rustc_middle::hir::map::collector::NodeCollector as rustc_hir::intravisit::Visitor>::visit_item::hcf243774ab523eaf
  34:     0x7fc83f23a2ea - rustc_middle::hir::map::index_hir::h040004cef163319f
  35:     0x7fc83f68e0a5 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h7d6682da11a84992
  36:     0x7fc83f264fe4 - rustc_data_structures::stack::ensure_sufficient_stack::h4cab318744a74eb3
  37:     0x7fc83f06e62c - rustc_query_system::query::plumbing::get_query_impl::ha9e8cde79fac1c1f
  38:     0x7fc83f285dad - core::ops::function::FnOnce::call_once::h7e89c6f7e4d763eb
  39:     0x7fc83f6a9935 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hf1400eebe2695697
  40:     0x7fc83f26d424 - rustc_data_structures::stack::ensure_sufficient_stack::h8fc56bf12b14bf52
  41:     0x7fc83f026543 - rustc_query_system::query::plumbing::get_query_impl::h48ba41acd58faf41
  42:     0x7fc83f2376b3 - rustc_middle::hir::map::Map::expect_item::h450057b4eb5544c0
  43:     0x7fc83d8d9ae5 - rustc_middle::hir::map::Map::visit_item_likes_in_module::h84da9a9ac99d9f0a
  44:     0x7fc83d866c3c - rustc_passes::hir_id_validator::check_crate::h87c8b1ff9806d76c
  45:     0x7fc83ca3dc8d - rustc_interface::passes::analysis::hf34e525a2b31f7c0
  46:     0x7fc83c777d7b - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute::h23cb841af51f0584
  47:     0x7fc83c816057 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h7178e9f1ddee6943
  48:     0x7fc83c8113df - rustc_middle::ty::query::plumbing::<impl rustc_query_system::query::QueryContext for rustc_middle::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}::{{closure}}::h4025824db8759ceb
  49:     0x7fc83c7a2f10 - rustc_query_system::query::plumbing::get_query_impl::h29928f066778d4c4
  50:     0x7fc83c77a12b - rustc_interface::passes::QueryContext::enter::h0705343d8e4a95be
  51:     0x7fc83c8134a2 - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter::h6eb07f11eba80bd6
  52:     0x7fc83c7da2a2 - rustc_span::with_source_map::h3820d41b3a036910
  53:     0x7fc83c8149c2 - rustc_interface::interface::create_compiler_and_run::h08ad396f4e7721e0
  54:     0x7fc83c7fcd5a - scoped_tls::ScopedKey<T>::set::hd425a36fbe06985e
  55:     0x7fc83c8199b5 - std::sys_common::backtrace::__rust_begin_short_backtrace::h23925571219dbc63
  56:     0x7fc83c7973ce - core::ops::function::FnOnce::call_once{{vtable.shim}}::h43b61b9af30b3669
  57:     0x7fc83bf3f21a - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h88864ee659ddb161
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/alloc/src/boxed.rs:1042:9
  58:     0x7fc83bf3f21a - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h9f03cbdc7ab33431
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/alloc/src/boxed.rs:1042:9
  59:     0x7fc83bf3f21a - std::sys::unix::thread::Thread::new::thread_start::h5979dd5560a11418
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys/unix/thread.rs:89:17
  60:     0x7fc83be243e9 - start_thread
  61:     0x7fc83bc0c293 - __GI___clone
  62:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.0.212 (ffa2e7a 2020-10-24)

query stack during panic:
#0 [index_hir] index HIR
#1 [hir_owner] HIR owner of `{misc#0}`
#2 [analysis] running analysis passes on this crate
end of query stack
warning: 1 warning emitted
  ```
  
  </p>
</details>

**Comment by ebroto** [comment]
Does not seem to reproduce in the [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=bd6a3fd7adb58f10a9c318db5e2eabbb) for the current Clippy.
